shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/NearbyHierarchy.kt

```diff
@@ -0,0 +1,433 @@
+
+private typealias MiddleLayerData = RouteAtStop<List<Direction>, InnerLayer>
+
+private typealias MiddleLayer = Map<Stop, MiddleLayerData>
+
+private typealias OuterLayer = Map<LineOrRoute, MiddleLayer>
+
+data class NearbyHierarchy(private val data: OuterLayer) {
+    sealed interface LineOrRoute {
+        data class Line(
+            val line: com.mbta.tid.mbta_app.model.Line,
+            val routes: Set<com.mbta.tid.mbta_app.model.Route>
+        ) : LineOrRoute
+
+        data class Route(val route: com.mbta.tid.mbta_app.model.Route) : LineOrRoute
+    }
+
+    data class RouteAtStop<StopData, T>(val stopData: StopData, val data: T) {
+        fun <U> map(f: (T) -> U): RouteAtStop<StopData, U> = RouteAtStop(stopData, f(data))
+    }
+
+    sealed interface DirectionOrHeadsign {
+        data class Direction(val direction: com.mbta.tid.mbta_app.model.Direction) :
+            DirectionOrHeadsign
+
+        data class Headsign(val headsign: String, val route: Route, val line: Line?) :
+            DirectionOrHeadsign
+    }
+
+    data class PartialHierarchy<StopData, T>(
+        val data: Map<LineOrRoute, Map<Stop, RouteAtStop<StopData, T>>>
+    ) {
+        fun <U> map(f: (T) -> U): PartialHierarchy<StopData, U> =
+            data
+                .mapValues { (_, routeData) ->
+                    routeData.mapValues { (_, routeAtStop) -> routeAtStop.map(f) }
+                }
+                .let(::PartialHierarchy)
+
+        fun <SD2, U> map(
+            processStopData: (StopData) -> SD2,
+            processData: (T) -> U
+        ): PartialHierarchy<SD2, U> =
+            data
+                .mapValues { (_, routeData) ->
+                    routeData.mapValues { (_, routeAtStop) ->
+                        RouteAtStop(
+                            processStopData(routeAtStop.stopData),
+                            processData(routeAtStop.data)
+                        )
+                    }
+                }
+                .let(::PartialHierarchy)
+
+        fun <U> mapInContext(
+            f: (LineOrRoute, Stop, StopData, T) -> U
+        ): PartialHierarchy<StopData, U> =
+            data
+                .mapValues { (lineOrRoute, routeData) ->
+                    routeData.mapValues { (stop, routeAtStop) ->
+                        routeAtStop.map { f(lineOrRoute, stop, routeAtStop.stopData, it) }
+                    }
+                }
+                .let(::PartialHierarchy)
+
+        fun <SD2> mapStopDataInContext(
+            f: (LineOrRoute, Stop, StopData, T) -> SD2
+        ): PartialHierarchy<SD2, T> =
+            data
+                .mapValues { (lineOrRoute, routeData) ->
+                    routeData.mapValues { (stop, routeAtStop) ->
+                        RouteAtStop(
+                            f(lineOrRoute, stop, routeAtStop.stopData, routeAtStop.data),
+                            routeAtStop.data
+                        )
+                    }
+                }
```

Comment:
suggestion: documentation for these various map functions would be helpful. The difference between these two is particularly confusing to me on first read, as is the significance of `SD2` as a generic param

Reply:
`U` is a natural successor to `T`, so the type argument for `mapInContext` was easy, but there's no obvious natural successor to `StopData`, so my inelegant compromise is `SD2`. If you think `StopData2` would be clearer, or some different name for `StopData` would be helpful, I can change those around.